### PR TITLE
registerProtocolHandler(): manual tests for percent encoding

### PR DESCRIPTION
**Pull request by annevk**
For https://github.com/whatwg/html/pull/5524.

### CODE REVIEW

File: html/webappapis/system-state-and-capabilities/the-navigator-object/resources/handler-tools.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
+// These can be used in an environment that has these variables defined:
+// * type
+// * handler
+// * nonASCIIExectations
```
Comment (saschanaz) [inline html/webappapis/system-state-and-capabilities/the-navigator-object/resources/handler-tools.js:4]: This seems redundant and includes a typo 👀

File: html/webappapis/system-state-and-capabilities/the-navigator-object/protocol-handler-fragment-manual.https.html [hunk 2]
```diff
@@ -0,0 +1,18 @@
+<!doctype html>
+<meta charset=windows-1254>
+<title>registerProtocolHandler() and a handler with %s in the fragment</title>
+<script src=/resources/testharness.js></script>
+<script src=/resources/testharnessreport.js></script>
+<script src=/service-workers/service-worker/resources/test-helpers.sub.js></script>
+<script>
+// Configure expectations for individual test
+let type = "fragment";
+let handler = "?QES\u2020QEE#FES\u2020FEEPSS%sPSE";
```
Comment (saschanaz) [inline html/webappapis/system-state-and-capabilities/the-navigator-object/protocol-handler-fragment-manual.https.html:10]: Maybe `const` as these never change.

**Review by saschanaz — APPROVED** [review]
Just nits

File: html/webappapis/system-state-and-capabilities/the-navigator-object/resources/handler-tools.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
+// These can be used in an environment that has these variables defined:
```
Comment (domenic) [inline html/webappapis/system-state-and-capabilities/the-navigator-object/resources/handler-tools.js:1]:
```suggestion
// These can be used in an environment that has these global variables defined:
```

File: html/webappapis/system-state-and-capabilities/the-navigator-object/protocol-handler-query-manual.https.html [hunk 4]
```diff
@@ -0,0 +1,18 @@
+<!doctype html>
+<meta charset=windows-1254>
+<title>registerProtocolHandler() and a handler with %s in the query</title>
+<script src=/resources/testharness.js></script>
+<script src=/resources/testharnessreport.js></script>
+<script src=/service-workers/service-worker/resources/test-helpers.sub.js></script>
+<script>
+// Configure expectations for individual test
+const type = "query";
```
Comment (domenic) [inline html/webappapis/system-state-and-capabilities/the-navigator-object/protocol-handler-query-manual.https.html:9]:
```suggestion
window.type = "query";
```

Comment (domenic) [reply html/webappapis/system-state-and-capabilities/the-navigator-object/protocol-handler-query-manual.https.html:9]: Same for the other 5. I like my global variables to be explicit.

File: html/webappapis/system-state-and-capabilities/the-navigator-object/protocol-handler-fragment-manual.https.html [hunk 5]
```diff
@@ -0,0 +1,18 @@
+<!doctype html>
+<meta charset=windows-1254>
```
Comment (domenic) [inline html/webappapis/system-state-and-capabilities/the-navigator-object/protocol-handler-fragment-manual.https.html:2]: If this is intentional it'd be good to add a comment stating so.

**Review by domenic — APPROVED** [review]
LGTM with nits

File: html/webappapis/system-state-and-capabilities/the-navigator-object/protocol-handler-fragment-manual.https.html (hunk 5 above)
Comment (annevk) [reply html/webappapis/system-state-and-capabilities/the-navigator-object/protocol-handler-fragment-manual.https.html:2]: Yeah, it's to test the `registerProtocolHandler()` call.

**Comment by domenic** [comment]
Adding timeout=long would make it easier to run these without timing out.

When I run this on Chrome on Windows I get a 404 when it tries to open a popup to

```
https://web-platform.test:8443/html/webappapis/system-state-and-capabilities/the-navigator-object/resources/handler/fragment/?QES%86QEE#FES%E2%80%A0FEEPSSweb%2Bwptfragment%3A%2501%2502%2503%2504%2505%2506%2507%2508%250B%250C%250E%250F%2510%2511%2512%2513%2514%2515%2516%2517%2518%2519%251A%251B%251C%251D%251E%251F+!%22%23%24%25%26'()*%2B%2C-.%2F0123456789%3A%3B%3C%3D%3E%3F%40ABCDEFGHIJKLMNOPQRSTUVWXYZ%5B%5C%5D%5E_%60abcdefghijklmnopqrstuvwxyz%7B%7C%7D~%7F%25C2%2580%25C2%2581PSE
```

**Comment by domenic** [comment]
In Firefox the test completes, but with an assert failure:

```
assert_equals: actual test expected "web%2Bwptfragment%3A%2501%2502%2503%2504%2505%2506%2507%2508%250B%250C%250E%250F%2510%2511%2512%2513%2514%2515%2516%2517%2518%2519%251A%251B%251C%251D%251E%251F%20!%22%23%24%25%26'()*%2B%2C-.%2F0123456789%3A%3B%3C%3D%3E%3F%40ABCDEFGHIJKLMNOPQRSTUVWXYZ%5B%5C%5D%5E_%60abcdefghijklmnopqrstuvwxyz%7B%7C%7D~%257F%25C2%2580%25C2%2581" but got "web%2Bwptfragment%3A%2501%2502%2503%2504%2505%2506%2507%2508%250B%250C%250E%250F%2510%2511%2512%2513%2514%2515%2516%2517%2518%2519%251A%251B%251C%251D%251E%251F%20!%2522%23%24%25%26'()*%2B%2C-.%2F0123456789%3A%3B%3C%3D%3E%3F%40ABCDEFGHIJKLMNOPQRSTUVWXYZ%5B%5C%5D%5E_%2560abcdefghijklmnopqrstuvwxyz%7B%7C%7D~%257F%25C2%2580%25C2%2581"
```

**Comment by domenic** [comment]
I wonder if protocol handler fetches bypass service workers in Chrome in general :O

**Comment by domenic** [comment]
Yes, they do :(. https://bugs.chromium.org/p/chromium/issues/detail?id=522370

Would it be possible to use a Python script handler instead? I realize that might be trickier since you'd likely need to stuff some data into the query string...

**Comment by annevk** [comment]
Then you are not testing replacement in the path. We could do that for query/fragment, but should probably be separate tests as it seems having some service worker coverage is useful.

The only failures I encountered in Firefox are those described at https://github.com/whatwg/html/issues/3377#issuecomment-625102116. In particular that it escapes U+0022 (") and U+0060 (`) twice, which is also what you're observing. (And it special cases U+0000, which may or may not be something the specification recommends, which reminds me that we should probably remove that text as well.)

**Comment by annevk** [comment]
As tested before, Chrome U+0020 as + and because for a non-special URL they do not encode U+007F, that's another failure.

There is one other thing I need to account for, U+0027. Which ends up encoded in the query as the handler is a special URL so when parsing the handler that needs to be percent-encoded.

(Both path and fragment percent-encode sets are clean subsets of the component encode-set so they don't have a similar problem.)

**Review by domenic — COMMENTED** [review]
Test structure looks good and these run in Chrome. There I get

```
assert_equals: actual test expected "web%2Bwptfragment%3A%2501%2502%2503%2504%2505%2506%2507%2508%250B%250C%250E%250F%2510%2511%2512%2513%2514%2515%2516%2517%2518%2519%251A%251B%251C%251D%251E%251F%20!%22%23%24%25%26'()*%2B%2C-.%2F0123456789%3A%3B%3C%3D%3E%3F%40ABCDEFGHIJKLMNOPQRSTUVWXYZ%5B%5C%5D%5E_%60abcdefghijklmnopqrstuvwxyz%7B%7C%7D~%257F%25C2%2580%25C2%2581" but got "web%2Bwptfragment%3A%2501%2502%2503%2504%2505%2506%2507%2508%250B%250C%250E%250F%2510%2511%2512%2513%2514%2515%2516%2517%2518%2519%251A%251B%251C%251D%251E%251F+!%22%23%24%25%26'()*%2B%2C-.%2F0123456789%3A%3B%3C%3D%3E%3F%40ABCDEFGHIJKLMNOPQRSTUVWXYZ%5B%5C%5D%5E_%60abcdefghijklmnopqrstuvwxyz%7B%7C%7D~%7F%25C2%2580%25C2%2581"
```

Can you post (probably to the spec issue) a consolidated summary of all the browser divergences from the tests, and then we can rope in appropriate Chrome people to confirm that they'd like to fix them eventually?

**Comment by annevk** [comment]
https://github.com/whatwg/html/issues/3377#issuecomment-628122390

**Comment by fred-wang** [comment]
Just for the record, after rebasing an old patch for https://bugs.chromium.org/p/chromium/issues/detail?id=522370 I verified that the sw tests and the path one in particular pass in Chrome too.

**Pull request merged**